File: base/src/main/java/com/zhangteng/base/widget/ItemMoveTouchHelper.java
package com.zhangteng.base.widget;

import android.graphics.Color;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;

import com.zhangteng.base.adapter.HeaderOrFooterAdapter;
import com.zhangteng.base.base.BaseAdapter;

import java.util.Collections;
import java.util.List;

/**
 * Created by swing on 2018/5/7.
 */
public class ItemMoveTouchHelper extends ItemTouchHelper {
    /**
     * Creates an ItemTouchHelper that will work with the given Callback.
     * <p>
     * You can attach ItemTouchHelper to a RecyclerView via
     * {@link #attachToRecyclerView(RecyclerView)}. Upon attaching, it will add an item decoration,
     * an onItemTouchListener and a Child attach / detach listener to the RecyclerView.
     *
     * @param callback The Callback which controls the behavior of this touch helper.
     */
    public ItemMoveTouchHelper(Callback callback) {
        super(callback);
    }

    @Override
    public void attachToRecyclerView(@Nullable RecyclerView recyclerView) {
        super.attachToRecyclerView(recyclerView);
    }

    public static class MoveTouchCallback extends ItemTouchHelper.Callback {

        /**
         * 如果是列表布局的话则拖拽方向为DOWN和UP，如果是网格布局的话则是DOWN和UP和LEFT和RIGHT
         */
        @Override
        public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
            if (recyclerView.getLayoutManager() instanceof GridLayoutManager) {
                final int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN |
                        ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT;
                final int swipeFlags = 0;
                return makeMovementFlags(dragFlags, swipeFlags);
            } else {
                final int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
                final int swipeFlags = 0;
                return makeMovementFlags(dragFlags, swipeFlags);
            }
        }

        /**
         * 将正在拖拽的item和集合的item进行交换元素
         */
        @Override
        public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
            BaseAdapter baseAdapter = (BaseAdapter) recyclerView.getAdapter();
            List datas = baseAdapter.getData();
            //得到当拖拽的viewHolder的Position
            int fromPosition = viewHolder.getAdapterPosition();
            //拿到当前拖拽到的item的viewHolder
            int toPosition = target.getAdapterPosition();
            int headerCount = baseAdapter.isHasHeaderOrFooter() ? ((HeaderOrFooterAdapter) baseAdapter).getHeadersCount() : 0;
            if (fromPosition < toPosition) {
                for (int i = fromPosition - headerCount; i < toPosition - headerCount; i++) {
                    Collections.swap(datas, i, i + 1);
                }
            } else {
                for (int i = fromPosition - headerCount; i > toPosition - headerCount; i--) {
                    Collections.swap(datas, i, i - 1);
                }
            }
            baseAdapter.notifyItemMoved(fromPosition, toPosition);
            return true;
        }

        /**
         * 拖曳处理后的回调
         */
        @Override
        public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {

        }

        /**
         * 长按选中Item的时候开始调用
         *
         * @param viewHolder
         * @param actionState
         */
        @Override
        public void onSelectedChanged(RecyclerView.ViewHolder viewHolder, int actionState) {
            if (actionState != ItemTouchHelper.ACTION_STATE_IDLE) {
                viewHolder.itemView.setBackgroundColor(Color.LTGRAY);
            }
            super.onSelectedChanged(viewHolder, actionState);
        }

        /**
         * 手指松开的时候还原
         *
         * @param recyclerView
         * @param viewHolder
         */
        @Override
        public void clearView(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
            super.clearView(recyclerView, viewHolder);
            viewHolder.itemView.setBackgroundColor(0);
        }
    }
}

File: base/src/main/java/com/zhangteng/base/widget/ItemMoveTouchHelper.kt
package com.zhangteng.base.widget

import android.graphics.Color
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.zhangteng.base.adapter.HeaderOrFooterAdapter
import com.zhangteng.base.base.BaseAdapter
import java.util.*

/**
 * Created by swing on 2018/5/7.
 */
open class ItemMoveTouchHelper
/**
 * Creates an ItemTouchHelper that will work with the given Callback.
 *
 *
 * You can attach ItemTouchHelper to a RecyclerView via
 * [.attachToRecyclerView]. Upon attaching, it will add an item decoration,
 * an onItemTouchListener and a Child attach / detach listener to the RecyclerView.
 *
 * @param callback The Callback which controls the behavior of this touch helper.
 */
(callback: Callback) : ItemTouchHelper(callback) {
    override fun attachToRecyclerView(recyclerView: RecyclerView?) {
        super.attachToRecyclerView(recyclerView)
    }

    class MoveTouchCallback : Callback() {
        /**
         * 如果是列表布局的话则拖拽方向为DOWN和UP，如果是网格布局的话则是DOWN和UP和LEFT和RIGHT
         */
        override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
            return if (recyclerView.getLayoutManager() is GridLayoutManager) {
                val dragFlags = UP or DOWN or
                        LEFT or RIGHT
                val swipeFlags = 0
                makeMovementFlags(dragFlags, swipeFlags)
            } else {
                val dragFlags = UP or DOWN
                val swipeFlags = 0
                makeMovementFlags(dragFlags, swipeFlags)
            }
        }

        /**
         * 将正在拖拽的item和集合的item进行交换元素
         */
        override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
            val baseAdapter = recyclerView.adapter as BaseAdapter<*, *>? ?: return true
            val data = baseAdapter.data
            //得到当拖拽的viewHolder的Position
            val fromPosition = viewHolder.adapterPosition
            //拿到当前拖拽到的item的viewHolder
            val toPosition = target.adapterPosition
            val headerCount = if (baseAdapter.hasHeaderOrFooter) (baseAdapter as HeaderOrFooterAdapter<*>?)!!.getHeadersCount() else 0
            if (fromPosition < toPosition) {
                for (i in fromPosition - headerCount until toPosition - headerCount) {
                    Collections.swap(data, i, i + 1)
                }
            } else {
                for (i in fromPosition - headerCount downTo toPosition - headerCount + 1) {
                    Collections.swap(data, i, i - 1)
                }
            }
            baseAdapter.notifyItemMoved(fromPosition, toPosition)
            return true
        }

        /**
         * 拖曳处理后的回调
         */
        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {}

        /**
         * 长按选中Item的时候开始调用
         *
         * @param viewHolder
         * @param actionState
         */
        override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
            if (actionState != ACTION_STATE_IDLE) {
                viewHolder?.itemView?.setBackgroundColor(Color.LTGRAY)
            }
            super.onSelectedChanged(viewHolder, actionState)
        }

        /**
         * 手指松开的时候还原
         *
         * @param recyclerView
         * @param viewHolder
         */
        override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
            super.clearView(recyclerView, viewHolder)
            viewHolder.itemView.setBackgroundColor(0)
        }
    }
}